### ISSUE TITLE
Add ValueTuple type forward to CoreRT

### ISSUE THREAD

**Issue by jcouv** [dotnet/corefx#15176]
ValueTuple was implemented in CoreRT. ~~Once it ships (I'm not sure which release or timeframe that will be),~~ The corefx package will need to type forward to CoreRT.

FYI @jkotas

**Comment by karelz** [comment]
@weshaggard @jkotas I think it might be best to track it in CoreRT repo - tied to "next release" milestone. CoreFX repo does not have milestones for CoreRT release cycle ...
Mainly I want to make sure we don't forget.

**Comment by jkotas** [comment]
Large fraction of CoreRT-related issues that needs to be fixed in corefx repo are common between ProjectN/UWP project and CoreRT project. This is one of them.

@karelz I think that we need tags to mark corefx issues specific to ProjectN/UWP, and corefx issues common to both ProjectN/UWP and CoreRT. You should assume that anything labelled CoreRT is in the common to both ProjectN/UWP and CoreRT bucket because ProjectN/UWP builds on top of CoreRT for the most part.

There will be small fraction of corefx issues that are really specific to CoreRT console apps and not apply to ProjectN/UWP. I think it would make sense to track these in corefx repo as well because of it is best to track the issues in the same repo as where the relevant code lives. If you would prefer to track these elsewhere, I am ok with it. I expect that they will be relatively rare so it does not matter a whole lot.

**Comment by karelz** [comment]
We have uwp tag "os-windows-uwp". If we don't expect tons of CoreRT-only issues, I would suggest to use just that.
Once we have schedule & milestone in CoreFX repo for UWP .NET Standard 2.0, we can move all "os-windows-uwp" issues into that milestone (or at least triage them from that perspective).

Another option is to create more labels / milestones and explain it to everyone - which IMO creates way too much complexity for something with little value (at least at this moment).

Does it sound reasonable?

**Comment by jkotas** [comment]
So the issue is tagged with `os-windows-uwp` that the code referrers to it as `UAP`, but it is really `corert`. Reminds me of [this joke](http://www.nejvtipy.cz/dva-policajti-zastavi-ridice-nakladniho-vozu_vt4115&skup=vtipy-o-policajtech) :-)

More seriously: I think we need a label for CoreRT so that you do not have to use full text search to look for it that never works too well: https://github.com/dotnet/corefx/search?utf8=%E2%9C%93&q=%5BCoreRT%5D&type=Issues

**Comment by karelz** [comment]
I am not against creating another label, if it brings value and if it can be easily and naturally explained to everyone :) (CoreRT vs. ProjectN/UWP is still a mystery to many ;-))
Let's chat and find out what is best.

**Comment by jkotas** [comment]
The rules should be:

- Does this issue affect Windows store apps? Use the existing os-windows-uwp label.
- Does this issue affect Console apps built using github.com/dotnet/corert/ toolchain? Use the new CoreRT label, potentially in combination with existing OS labels if it is OS specific.
- Does it affect both? Use both labels.

> CoreRT vs. ProjectN/UWP is still a mystery to many

https://github.com/dotnet/corert/blob/master/Documentation/intro-to-corert.md and https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/glossary.md are the two documents that touch on this. Is there something that needs to get fixed up in these docs to make the difference easier to understand?

**Comment by joperezr** [comment]
@jcouv @jkotas why do we have to wait for CoreRT to ship in order to do the required work on corefx side? can't we just use the NETNative TargetingPack with the latest changes and type forward to that? If not, then what am I missing?

**Comment by joperezr** [comment]
I agree with the idea of the new label in order to separate CoreRT specific issues. @karelz do you agree.

**Comment by jkotas** [comment]
> why do we have to wait for CoreRT to ship in order to do the required work on corefx side

We do not. I do think that anything in this issue says that it is the case.

**Comment by joperezr** [comment]
I see, well the description of this issue says
> Once it ships (I'm not sure which release or timeframe that will be), 

Which led me to believe that we needed an official package shipped in nuget.org in order to do the work here. @jcouv do you still want to take this or should I just flag it as up-for-grabs?

**Comment by karelz** [comment]
I had a chat about it with @jkotas about the CoreRT label. We didn't come to full agreement yet.

My latest proposal:
* Add os-corert (in theory it should be platform-*, but it is at least consistent now with os-windows-uwp and os-windows-wsl)
  * Note: We can consider renaming all os-* to platform-* as independent change (I am not against that)
* Use os-corert primarily on issues which are CoreRT-specific and don't affect UWP.
  * Note: It's similar to os-linux and os-mac-os-x: We do not blast os-mac-os-x label rigorously on issues which already have os-linux ... the assumption is that os-linux will typically address both Linux and Mac. We use os-mac-os-x mostly on issues which are truly Mac specific.
  * Key value: I don't want to burden developers to understand full matrix of platforms and use labels like crazy, unless it is helping them or helping to track the project.

@jkotas are you ok with the plan? Or do you want to tweak it before we roll it out?

**Comment by jkotas** [comment]
I think that os- prefix for corert is *super confusing*.

**Comment by jkotas** [comment]
You may also add `platform-mono` and `plaform-netfx` because of there are quite a few of issues in corefx specific to those.

**Comment by jcouv** [comment]
@joperezr My ultimate goal is that the ValueTuple package (from CoreFX) needs to have type-forwards to account for the CoreRT implementation. I defer to the team in terms of timing and dependencies, 

My understanding was that such type-forwards are currently blocked on shipping CoreRT with ValueTuple and possibly also some other work. If that is not the case, please provide some details (what is the moniker I should use?) and assign the issue to me to go ahead and make the change.

**Comment by jkotas** [comment]
> the CoreRT implementation

The CoreRT implementation is exactly same as UAPAOT implementation, so the work is identical between the two.

>  currently blocked on shipping CoreRT with ValueTuple 

No, CoreRT and UAPAOT ship independently. However once the work is done, it does not matter which one ships first.

**Comment by karelz** [comment]
`platform-mono` might be eventually useful.

Which bugs should have `platform-netfx`? Overall we try to NOT track anything for Desktop/netfx on GitHub (netfx-port-consider is the only tag which triggers mirroring of issues into internal TFS for Desktop purposes).

**Comment by jkotas** [comment]
corefx repo is building a lot of bits that specifically targeting full framework platform (OOB assemblies, targeting packs, etc.) `platform-netfx` is yet another vertical, just like uap or corert verticals. Anything specific to this vertical would be marked with `platform-netfx`. A few examples:
- https://github.com/dotnet/corefx/issues/15954 
- https://github.com/dotnet/corefx/issues/14101
- https://github.com/dotnet/corefx/issues/9045
- https://github.com/dotnet/corefx/pull/15344
- https://github.com/dotnet/corefx/pull/15664

**Comment by joperezr** [comment]
@jcouv @jkotas I checked and actually this work(type forwarding ValueTuple to CoreRT's CoreLib) is already done, given that System.Runtime was already compiling against corert in uapaot build. That means that we can close this issue.

@jkotas @karelz do you want to change the title of this issue to be about the label discussion? or is it ok to just close it?

**Comment by karelz** [comment]
@joperezr the label discussion should be somewhere else. Sorry for hijacking this issue. Feel free to close it.

**Comment by jcouv** [comment]
@joperezr I also tried to verify this is already done. I must be doing it wrong.

I downloaded the latest beta package (from [myget](https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.ValueTuple)) and opened it.
Both in the `ref` and `lib` folders I see 3 sub-folders: PCL, netstandard1.0 and netcoreapp1.1.
But I would expect a fourth sub-folder for a moniker corresponding to CoreRT.
